core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java

```diff
@@ -72,34 +78,55 @@ protected ClusterBlockException checkBlock(GetSnapshotsRequest request, ClusterS
     }
 
     @Override
-    protected void masterOperation(final GetSnapshotsRequest request, ClusterState state, final ActionListener<GetSnapshotsResponse> listener) {
+    protected void masterOperation(final GetSnapshotsRequest request, ClusterState state,
+                                   final ActionListener<GetSnapshotsResponse> listener) {
         try {
+            final String repository = request.repository();
             List<SnapshotInfo> snapshotInfoBuilder = new ArrayList<>();
             if (isAllSnapshots(request.snapshots())) {
-                snapshotInfoBuilder.addAll(snapshotsService.snapshots(request.repository(), request.ignoreUnavailable()));
+                snapshotInfoBuilder.addAll(snapshotsService.snapshots(repository, snapshotsService.snapshotIds(repository),
+                    request.ignoreUnavailable()));
             } else if (isCurrentSnapshots(request.snapshots())) {
-                snapshotInfoBuilder.addAll(snapshotsService.currentSnapshots(request.repository()));
+                snapshotInfoBuilder.addAll(snapshotsService.currentSnapshots(repository));
             } else {
-                Set<String> snapshotsToGet = new LinkedHashSet<>(); // to keep insertion order
-                List<SnapshotInfo> snapshots = null;
+                final Set<SnapshotId> currentSnapshotIds = new LinkedHashSet<>(
+                    snapshotsService.currentSnapshots(repository).stream().map(SnapshotInfo::snapshotId).collect(Collectors.toList())
+                );
+                final Map<String, SnapshotId> allSnapshotIds = Stream.concat(
+                    currentSnapshotIds.stream(),
+                    snapshotsService.snapshotIds(repository).stream()
+                ).collect(Collectors.toMap(SnapshotId::getName, Function.identity(), (SnapshotId s1, SnapshotId s2) -> {
+                    // duplicates in the map - there is a small window of time where overlap is possible between the
+                    // repository writing the snapshot blob and the cluster service removing the in-progress snapshot
+                    // from its cluster state, so we will have duplicates in the map.
+                    if (currentSnapshotIds.contains(s1)) {
+                        return s2;
+                    } else {
+                        return s1;
+                    }
+                }));
```

Comment:
I agree, it became cumbersome b/c of having to handle the fact that `Collectors#toMap` requires a merge strategy for duplicate keys (hence the third lamda parameter).